File: examples/elm-core/ir-spec/map.ml
type ('tK, 'tV) map'2 = ('tK, 'tV) Map_ffi.map'2

let rec empty'0 : unit -> ('t_K, 't_V) map'2 = function
  | () -> Map_ffi.empty'0 ()

let rec get'2 : 'tK * ('tK, 'tV) map'2 -> 'tV Maybe.maybe'1 = function
  | v_Key, v_Map -> Map_ffi.get'2 (v_Key, v_Map)

let rec member'2 : 'tK * ('tK, 't_V) map'2 -> bool = function
  | v_Key, v_Map -> (
      match get'2 (v_Key, v_Map) with
      | Maybe'Just _ -> true
      | Maybe'Nothing -> false )

let rec size'1 : ('t_K, 't_V) map'2 -> int = function
  | v_Map -> Map_ffi.size'1 v_Map

let rec is_empty'1 : ('t_K, 't_V) map'2 -> bool = function
  | v_Map -> size'1 v_Map = 0

let rec insert'3 : 'tK * 'tV * ('tK, 'tV) map'2 -> ('tK, 'tV) map'2 = function
  | v_Key, v_Value, v_Map -> Map_ffi.insert'3 (v_Key, v_Value, v_Map)

let rec remove'2 : 'tK * ('tK, 'tV) map'2 -> ('tK, 'tV) map'2 = function
  | v_Key, v_Map -> Map_ffi.remove'2 (v_Key, v_Map)

let rec update'3 :
    'tK * ('tV Maybe.maybe'1 -> 'tV Maybe.maybe'1) * ('tK, 'tV) map'2 ->
    ('tK, 'tV) map'2 = function
  | v_Key, v_Alter, v_Map -> (
      match v_Alter (get'2 (v_Key, v_Map)) with
      | Maybe'Just v_Value -> insert'3 (v_Key, v_Value, v_Map)
      | Maybe'Nothing -> remove'2 (v_Key, v_Map) )

let rec singleton'2 : 'tK * 'tV -> ('tK, 'tV) map'2 = function
  | v_Key, v_Value -> insert'3 (v_Key, v_Value, empty'0 ())

let rec union'2 : ('tK, 'tV) map'2 * ('tK, 'tV) map'2 -> ('tK, 'tV) map'2 =
  function
  | v_Map1, v_Map2 -> Map_ffi.union'2 (v_Map1, v_Map2)

let rec fold'3 : ('tK * 'tV * 'tR -> 'tR) * 'tR * ('tK, 'tV) map'2 -> 'tR =
  function
  | v_F, v_Acc, v_Map -> Map_ffi.fold'3 (v_F, v_Acc, v_Map)

let rec map'2 : ('tK * 'tA -> 'tB) * ('tK, 'tA) map'2 -> ('tK, 'tB) map'2 =
  function
  | v_F, v_Map -> Map_ffi.map'2 (v_F, v_Map)

let rec filter'2 : ('tK * 'tV -> bool) * ('tK, 'tV) map'2 -> ('tK, 'tV) map'2 =
  function
  | v_F, v_Map -> Map_ffi.filter'2 (v_F, v_Map)

let rec intersect'2 : ('tK, 'tV) map'2 * ('tK, 'tV) map'2 -> ('tK, 'tV) map'2 =
  function
  | v_Map1, v_Map2 ->
      filter'2 ((function v_K, _ -> member'2 (v_K, v_Map2)), v_Map1)

let rec diff'2 : ('tK, 'tV) map'2 * ('tK, 'tV) map'2 -> ('tK, 'tV) map'2 =
  function
  | v_Map1, v_Map2 ->
      filter'2 ((function v_K, _ -> not (member'2 (v_K, v_Map2))), v_Map1)

let rec partition'2 :
    ('tK * 'tV -> bool) * ('tK, 'tV) map'2 ->
    ('tK, 'tV) map'2 * ('tK, 'tV) map'2 = function
  | v_F, v_Map ->
      let v_Add = function
        | v_K, v_V, (v_D1, v_D2) -> (
            match v_F (v_K, v_V) with
            | true -> (insert'3 (v_K, v_V, v_D1), v_D2)
            | false -> (v_D1, insert'3 (v_K, v_V, v_D2)) )
      in
      fold'3 (v_Add, (empty'0 (), empty'0 ()), v_Map)

let rec keys'1 : ('tK, 't_V) map'2 -> 'tK list = function
  | v_Map -> Map_ffi.keys'1 v_Map

let rec values'1 : ('t_K, 'tV) map'2 -> 'tV list = function
  | v_Map -> Map_ffi.values'1 v_Map

let rec to_list'1 : ('tK, 'tV) map'2 -> ('tK * 'tV) list = function
  | v_Map -> Map_ffi.to_list'1 v_Map

let rec from_list'1 : ('tK * 'tV) list -> ('tK, 'tV) map'2 = function
  | v_List -> Map_ffi.from_list'1 v_List

File: examples/elm_core/ir-spec/map.erl
-file("elm_core/src/map.erlt", 1).

-module(map).

-export_type([map/2]).

-export([empty/0,
         get/2,
         member/2,
         size/1,
         is_empty/1,
         insert/3,
         remove/2,
         update/3,
         singleton/2]).

-export([union/2, intersect/2, diff/2]).

-export([fold/3, map/2, filter/2, partition/2]).

-export([keys/1, values/1, to_list/1, from_list/1]).

-import_type({maybe, [{maybe, 1}]}).

-opaque map(K, V) :: map_ffi:map(K, V).

-spec empty() -> map(_K, _V).

empty() -> map_ffi:empty().

-spec get(K, map(K, V)) -> maybe:maybe(V).

get(Key, Map) -> map_ffi:get(Key, Map).

-spec member(K, map(K, _V)) -> boolean().

member(Key, Map) ->
    case get(Key, Map) of
        {969696, maybe, maybe, just, _} -> true;
        {969696, maybe, maybe, nothing} -> false
    end.

-spec size(map(_K, _V)) -> integer().

size(Map) -> map_ffi:size(Map).

-spec is_empty(map(_K, _V)) -> boolean().

is_empty(Map) -> map:size(Map) =:= 0.

-spec insert(K, V, map(K, V)) -> map(K, V).

insert(Key, Value, Map) ->
    map_ffi:insert(Key, Value, Map).

-spec remove(K, map(K, V)) -> map(K, V).

remove(Key, Map) -> map_ffi:remove(Key, Map).

-spec update(K, fun((maybe:maybe(V)) -> maybe:maybe(V)),
             map(K, V)) -> map(K, V).

update(Key, Alter, Map) ->
    case Alter(get(Key, Map)) of
        {969696, maybe, maybe, just, Value} ->
            insert(Key, Value, Map);
        {969696, maybe, maybe, nothing} -> remove(Key, Map)
    end.

-spec singleton(K, V) -> map(K, V).

singleton(Key, Value) -> insert(Key, Value, empty()).

-spec union(map(K, V), map(K, V)) -> map(K, V).

union(Map1, Map2) -> map_ffi:union(Map1, Map2).

-spec intersect(map(K, V), map(K, V)) -> map(K, V).

intersect(Map1, Map2) ->
    filter(fun (K, _) -> member(K, Map2) end, Map1).

-spec diff(map(K, V), map(K, V)) -> map(K, V).

diff(Map1, Map2) ->
    filter(fun (K, _) -> not member(K, Map2) end, Map1).

-spec fold(fun((K, V, R) -> R), R, map(K, V)) -> R.

fold(F, Acc, Map) -> map_ffi:fold(F, Acc, Map).

-spec map(fun((K, A) -> B), map(K, A)) -> map(K, B).

map(F, Map) -> map_ffi:map(F, Map).

-spec filter(fun((K, V) -> boolean()),
             map(K, V)) -> map(K, V).

filter(F, Map) -> map_ffi:filter(F, Map).

-spec partition(fun((K, V) -> boolean()),
                map(K, V)) -> {map(K, V), map(K, V)}.

partition(F, Map) ->
    Add = fun (K, V, {D1, D2}) ->
                  case F(K, V) of
                      true -> {insert(K, V, D1), D2};
                      false -> {D1, insert(K, V, D2)}
                  end
          end,
    fold(Add, {empty(), empty()}, Map).

-spec keys(map(K, _V)) -> [K].

keys(Map) -> map_ffi:keys(Map).

-spec values(map(_K, V)) -> [V].

values(Map) -> map_ffi:values(Map).

-spec to_list(map(K, V)) -> [{K, V}].

to_list(Map) -> map_ffi:to_list(Map).

-spec from_list([{K, V}]) -> map(K, V).

from_list(List) -> map_ffi:from_list(List).



